integration/upload_test.go

```diff
@@ -0,0 +1,229 @@
+
+	if token := mqttClient.Connect(); token.Wait() && token.Error() != nil {
+		require.NoError(suite.T(), token.Error(), "connect to MQTT broker")
+	}
+
+	thingCfg, err := getThingConfig(mqttClient)
+	if err != nil {
+		mqttClient.Disconnect(uint(cfg.MqttQuiesceMs))
+		require.NoError(suite.T(), err, "get thing config")
+	}
+
+	suite.T().Logf("thing config: %+v", *thingCfg)
+
+	dittoClient, err := ditto.NewClientMqtt(mqttClient, ditto.NewConfiguration())
+	if err == nil {
+		err = dittoClient.Connect()
+	}
+
+	if err != nil {
+		mqttClient.Disconnect(uint(cfg.MqttQuiesceMs))
+		require.NoError(suite.T(), err, "initialize ditto client")
+	}
+
+	feature := &model.Feature{}
+	cmd := things.NewCommand(model.NewNamespacedIDFrom(thingCfg.DeviceID)).Twin().Feature(featureID).
+		Modify(feature)
+	msg := cmd.Envelope(protocol.WithResponseRequired(false))
+
+	err = dittoClient.Send(msg)
+	require.NoError(suite.T(), err, "create test feature")
+
+	suite.dittoClient = dittoClient
+	suite.mqttClient = mqttClient
+	suite.cfg = cfg
+	suite.thingCfg = thingCfg
+
+	suite.thingURL = fmt.Sprintf("%s/api/2/things/%s", strings.TrimSuffix(cfg.DittoAddress, "/"), thingCfg.DeviceID)
+	suite.featureURL = fmt.Sprintf("%s/features/%s", suite.thingURL, featureID)
+
+	uploadDir, err = setupUploadDir()
+	require.Nil(suite.T(), err, "get configured file upload directory")
+	require.NotEmpty(suite.T(), uploadDir, "get configured file upload directory")
+	suite.T().Logf("upload dir - %s", uploadDir)
+}
+
+func (suite *uploadSuite) TearDownSuite() {
+	suite.dittoClient.Disconnect()
+	suite.mqttClient.Disconnect(uint(suite.cfg.MqttQuiesceMs))
+}
+
+func setupUploadDir() (string, error) {
+	var uploadDir string
+	data, err := ioutil.ReadFile(configFile)
+	if err != nil {
+		return uploadDir, err
+	}
+	config := make(map[string]interface{})
+	err = json.Unmarshal(data, &config)
+	if err != nil {
+		return uploadDir, err
+	}
+	if files, ok := config["files"]; ok {
+		uploadDir = filepath.Dir(files.(string))
+	}
+	return uploadDir, nil
+}
+
+func TestFileUpload(t *testing.T) {
+	suite.Run(t, new(uploadSuite))
+}
+
+func (suite *uploadSuite) TestUploadHTTP() {
+	suite.T().Log("test file upload over HTTP")
+	uploadHandler := newHTTPUploadHandler()
+	suite.testUpload(uploadHandler)
+}
+
+func (suite *uploadSuite) testUpload(uploadHandler uploadHandler) {
+	files, err := createTestFiles(uploadDir)
+	defer deleteTestFiles(files)
+	require.Nil(suite.T(), err, "create test files")
+
+	err = uploadHandler.prepare()
+	require.Nil(suite.T(), err, "prepare upload handler")
+	defer uploadHandler.dispose()
+
+	type uploadRequest struct {
+		CorrelationID string            `json:"correlationId"`
+		Options       map[string]string `json:"options"`
+	}
+	requestPath := fmt.Sprintf("/features/%s/outbox/messages/request", featureID)
+	filePaths := make(map[string]string)
+	eventTopic := "e"
+	suite.mqttClient.Subscribe(eventTopic, 1, func(c mqtt.Client, m mqtt.Message) {
```

Comment:
The internal MQTT communication should not be used by this TC. The event can be received from Ditto via WebSocket.